packages/ubuntu_wizard/lib/src/wizard_button.dart

```diff
@@ -152,52 +70,206 @@ class WizardButton extends StatefulWidget {
+      child: showSpinner
+          ? SizedBox.square(
+              dimension: IconTheme.of(context).size,
+              child: const YaruCircularProgressIndicator(strokeWidth: 3),
+            )
+          : Text(widget.label!),
+    );
+  }
+
+  void _startLoadingTimer() {
+    _cancelTimer();
+    _loadingTimer = Timer(_kLoadingDelay, () {
+      if (mounted && loading) {
+        setState(() => showSpinner = true);
+      }
+    });
+  }
+
+  void _cancelTimer() {
+    _loadingTimer?.cancel();
+    _loadingTimer = null;
+  }
+
+  @override
+  void dispose() {
+    _loadingTimer?.cancel();
+    super.dispose();
+  }
+}
+
+/// A _Previous_ button that returns back to the previous page.
+class PreviousWizardButton extends StatelessWidget {
+  const PreviousWizardButton({
+    super.key,
+    this.visible = true,
+    this.enabled,
+    this.onBack,
+  });
+
+  final bool visible;
+  final bool? enabled;
+  final WizardCallback? onBack;
+
+  @override
+  Widget build(BuildContext context) {
+    final wizard = Wizard.maybeOf(context);
+    final rootWizard = Wizard.maybeOf(context, root: true);
+    final routeData =
+        (wizard?.routeData ?? rootWizard?.routeData) as WizardRouteData?;
+    final hasPrevious = [
+      routeData?.hasPrevious,
+      wizard?.hasPrevious,
+      rootWizard?.hasPrevious,
+    ].contains(true);
+    final isLoading = [wizard?.isLoading, rootWizard?.isLoading].contains(true);
+
+    return AnimatedBuilder(
+      animation: wizard?.controller ?? _noAnimation,
+      builder: (context, child) => WizardButton(
+          label: UbuntuLocalizations.of(context).previousLabel,
+          visible: visible,
+          flat: true,
+          enabled: !isLoading && (enabled ?? hasPrevious),
+          onActivated: onBack,
+          execute: () {
+            // navigate the root wizard at the end of a nested wizard
+            final effectiveWizard =
+                (wizard?.hasPrevious ?? false) ? wizard : rootWizard;
+            return effectiveWizard?.back();
+          }),
+    );
+  }
+}
+
+/// A _Next_ button that advances to the next page.
+class NextWizardButton extends StatelessWidget {
+  const NextWizardButton({
+    super.key,
+    this.label,
+    this.visible = true,
+    this.enabled = true,
+    this.flat = false,
+    this.highlighted = false,
+    this.arguments,
+    this.onNext,
+    this.onExecute,
+  });
+
+  final String? label;
+  final bool visible;
+  final bool enabled;
+  final bool flat;
+  final bool highlighted;
+  final Object? arguments;
+  final WizardCallback? onNext;
+  final WizardCallback? onExecute;
+
+  @override
+  Widget build(BuildContext context) {
+    final wizard = Wizard.maybeOf(context);
+    final rootWizard = Wizard.maybeOf(context, root: true);
+    final routeData =
+        (wizard?.routeData ?? rootWizard?.routeData) as WizardRouteData?;
+    final hasNext = [
+      routeData?.hasNext,
+      wizard?.hasNext,
+      rootWizard?.hasNext,
```

Comment:
It seems like these are all eventually non-null, do you know if this is always the case @d-loose?

I don't understand why the previous button implementation didn't show the done label before, because if I copy in that logic here now it still shows the done label. Anyhow, this seems to work like it should.

Reply:
Ah, `routeData` can of course be null sometimes, fixed that.

Reply:
Right, `routeData` is entirely optional.
`wizard` and `rootWizard` are only nullable because `Wizard.maybeOf` is used instead of `Wizard.of` here. Not sure if there's a sensible case where either of those can be null.

Reply:
I'm just realizing that this broke #340. `routeData` should always take precedence if provided, otherwise it doesn't serve any purpose.